text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,542 @@
+which people were previously required to use.
+
+## Deprecations
+
+ * Deprecate `std::process::Command::output()`.
+   This API cannot be fixed;
+   see [#73126](https://github.com/rust-lang/rust/issues/73126).
+
+ * Apply `#[must_use]` to `Command` and `Child`.
+
+ * Apply `#[must_use]` to `ExitStatus`.
+   (May require fixing quite a few of the examples.)
+
+## stderr handling
+
+If `stderr(Stdio::piped())`,
+these new functions all collect the child's stderr.
+Then,
+if the stderr output is nonempty, this is considered an error,
+and reported in the `SubprocessError`.
+
+These functions *do not* wait for EOF on stderr.
+Rather, they wait for child process termination and expect that
+any relevant error messages have been printed by that point.
+Any further stderr output 
+(for example, from forked but unawaited children of the command)
+might not be reported via the Rust std API.
+Such output might be ignored,
+or might accumulate in a temporary disk file not deleted
+until the last escaped handle onto the file has gone.
+Ideally, any escaped writer(s) would experience broken pipe errors.
+
+This non-waiting behaviour on stderr could be important
+when Rust programs invoke tools like `ssh`,
+which sometimes hang onto their stderr well after the
+intended remote command has completed.
+
+The implementation may involve a temporary file,
+or an in-memory buffer,
+or both.
+
+## New `struct SubprocessError`
+
+This new struct is used as the error type for the new methods.
+
+It can represents zero or more of the various
+distinct problems that can occur while running a process.
+A `SubprocessError` returned by a `std` function will always
+represent at least one problem (unless otherwise stated),
+but it may represent several
+
+For example a process which exited nonzero
+probably printed to stderr;
+with `piped` we capture that, and represent both the stderr
+and the exit status as problems within the `SubprocessError`.
+
+```rust
+impl SubprocessError {
+    /// The program, if we know it.
+    fn program(&self) -> Option<&OsStr>;
+
+    /// The arguments, if we know them.
+    fn args(&self) -> Option<std::process::CommandArgs<'_>>;
+
+    /// If the stdout was captured in memory, the stdout data.
+    fn stdout_bytes(&self) -> Option<&[u8]>;
+
+    /// If the process exited and we collected its status, the exit status.
+    fn status(&self) -> Option<ExitStatus>;
+
+    /// If trouble included nonempty stderr, the captured stderr
+    fn stderr_bytes(&self) -> Option<&[u8]>;
+
+    /// If trouble included failure to spawn, the spawn error.
+    fn spawn_error() -> Option<&io::Error>;
+
+    /// If trouble included failure to talk to the child, the IO error.
+    ///
+    /// This might include problems which might be caused by child
+    /// misbehaviour.
+    fn communication_error() -> Option<&io::Error>;
+
+    /// If trouble included failed UTF-8 conversion.
+    fn utf8_error(&self) -> Option<&std::str::FromUtf8Error>;
```

Comment:
Which UTF-8 conversion does this refer to? I think this probably should be pushed down to whatever getter would return something UTF-8 validated rather than having this stored as a field on the error.

If this *is* stored (and almost certainly if it is not), we should use `std::std::Utf8Error` rather than a `&std::string::FromUtf8Error` in order to avoid needing to store the invalid UTF8 twice (and to avoid forcing us to convert eagerly).

Reply:
I'm using `std::str::FromUtf8Error` which doesn't contain or reference the string, but does contain the offset.  I think this doesn't suffer from the problem you describe?